src/bios.rs

```diff
@@ -0,0 +1,145 @@
+            bail!("Failed to find {:?}", grub_install);
+        }
+
+        let boot_dir = Path::new(dest_root).join("boot");
+
+        #[cfg(target_arch = "x86_64")]
+        let r = std::process::Command::new(grub_install)
+            .args(&["--target", "i386-pc"])
+            .args(&["--boot-directory", boot_dir.to_str().unwrap()])
+            .args(&["--modules", "mdraid1x"])
+            .arg(&device)
+            .status()?;
+
+        #[cfg(target_arch = "powerpc64")]
+        let r = std::process::Command::new(grub_install)
+            .args(&["--target", "powerpc-ieee1275"])
+            .args(&["--boot-directory", boot_dir.to_str().unwrap()])
+            .arg("--no-nvram")
+            .arg(&device)
+            .status()?;
+
+        if !r.success() {
+            anyhow::bail!("Failed to run grub2-install");
+        }
+        Ok(InstalledContent {
+            meta,
+            filetree: None,
+            adopted_from: None,
+        })
+    }
+
+    fn generate_update_metadata(&self, sysroot_path: &str) -> Result<ContentMetadata> {
+        let grub_install = Path::new(sysroot_path).join(GRUB_BIN);
+        if !grub_install.exists() {
+            bail!("Failed to find {:?}", grub_install);
+        }
+
+        // Query the rpm database and list the package and build times for /usr/sbin/grub2-install
+        let mut rpmout = util::rpm_query(sysroot_path, &grub_install)?;
+        let rpmout = rpmout.output()?;
+        if !rpmout.status.success() {
+            std::io::stderr().write_all(&rpmout.stderr)?;
+            bail!("Failed to invoke rpm -qf");
+        }
+
+        let meta = util::parse_rpm_metadata(rpmout.stdout)?;
+        write_update_metadata(sysroot_path, self, &meta)?;
+        Ok(meta)
+    }
+
+    fn query_update(&self, sysroot: &openat::Dir) -> Result<Option<ContentMetadata>> {
+        get_component_update(sysroot, self)
+    }
+
+    #[allow(unused_variables)]
+    fn query_adopt(&self) -> Result<Option<Adoptable>> {
+        todo!();
+    }
+
+    #[allow(unused_variables)]
+    fn adopt_update(
+        &self,
+        sysroot: &openat::Dir,
+        update: &ContentMetadata,
+    ) -> Result<InstalledContent> {
+        todo!();
+    }
+
+    #[allow(unused_variables)]
+    fn run_update(
+        &self,
+        sysroot: &openat::Dir,
+        _: &InstalledContent,
+    ) -> Result<InstalledContent> {
+        let updatemeta = self.query_update(sysroot)?.expect("update available");
+        
+        let grub_install = Path::new("/").join(GRUB_BIN);
+        if !grub_install.exists() {
+            bail!("Failed to find {:?}", grub_install);
+        }
+
+        let boot_dir = Path::new("/").join("boot");
+
+        #[cfg(target_arch = "x86_64")]
+        let r = std::process::Command::new(grub_install)
+            .args(&["--target", "i386-pc"])
+            .args(&["--boot-directory", boot_dir.to_str().unwrap()])
+            .args(&["--modules", "mdraid1x"])
+            .arg(&device)
```

Comment:
Not sure how to get `device` name for update, any suggestions?

Reply:
Good question!  We could look at the backing device of `/boot` (or `/`) as a default.  (For example, forking off `findmnt -J /boot`, then we need to find its parent device.  There's a lot of related code in https://github.com/coreos/coreos-installer/blob/main/src/bin/rdcore/rootmap.rs

Reply:
Good suggestion, will look at it, thanks!